flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkResourceContext.java

```diff
@@ -65,7 +65,7 @@ public KubernetesJobAutoScalerContext getJobAutoScalerContext() {
     }
 
     private KubernetesJobAutoScalerContext createJobAutoScalerContext() {
-        Configuration conf = new Configuration(getObserveConfig());
+        Configuration conf = new Configuration(getDeployConfig(resource.getSpec()));
```

Comment:
@gyfora @1996fanrui I changed this logic because I believe it is the right config to return in the context of the autoscaler. We want to run autoscaling based on the current user-provided configuration, not based on what is currently deployed. The former is the user intend, the latter is the result of a past reconciliation decision. This will also ensure that overrides are cleared immediately and not after completing the full reconciliation.

For the tuning, the current logic only works correctly if we operate on the "deploy config" because that allows us to correctly determine the original memory configuration which is required for dynamically increasing / decreasing total memory and managed memory without going past the original user-provided limits.

Reply:
I think using the observe config is the right thing to do for the autoscaler in general. We are observing a flink job with its metrics and configuration and making scaling decisions.

The deploy config may contain yet undeployed settings .

I think this change only makes this feature more convenient to implement but I think the right thing to do would be to instead store the original memory limit in the autoscaler if that's important for the logic.

Reply:
Also for the standalone autoscaler you may only have the config that is actually running on the cluster (again that is equivalent to the observe config here)

Reply:
Storing memory limits or user-specified configuration wouldn't be correct because those always come from the spec, but I can add a getter to the Autoscaler context to retrieve the memory limit from the spec. That's what I initially had.

Reply:
But you are right that if the user actually changes the memory limits in a new version of the spec then the deploy config would actually make more sense to use.

Reply:
Thinking through this again, I think you are right, that in some cases the deploy config is simply better. Probably the only clear solution is to provide both but we could also start with just the deploy config